<a href="https://colab.research.google.com/github/sbarrerab/AlgorithmsUN2023II/blob/main/Lab4/sbarrerabGroup02Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.172
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4400.34
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [56]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

5 7
12

### Test input output files

In [93]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [94]:
%%writefile output1.txt
12

Writing output1.txt


In [95]:
%%writefile input2.txt
0 9

Writing input2.txt


In [96]:
%%writefile output2.txt
9

Writing output2.txt


In [97]:
!./APlusB < input1.txt

/bin/bash: line 1: ./APlusB: No such file or directory


In [98]:
!./APlusB < input2.txt

/bin/bash: line 1: ./APlusB: No such file or directory


### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

4 7
11


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [58]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [59]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [60]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [61]:
%%writefile output0.txt
0

Writing output0.txt


In [62]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [99]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [64]:
!./max_pairwise_product_full < input1.txt

90


In [100]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [101]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [102]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [103]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [104]:
200000*200000

40000000000

In [105]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [2]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [106]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [30]:
%%writefile input4.txt
9
300000 41 300000 9 33 0 7 245 24569

Overwriting input4.txt


In [31]:
300000 * 300000

90000000000

In [32]:
%%writefile output4.txt
90000000000

Overwriting output4.txt


In [33]:
%%writefile input5.txt
8
8 0 300000 66 30000 8955 895 966

Overwriting input5.txt


In [34]:
300000 * 30000

9000000000

In [35]:
%%writefile output5.txt
9000000000

Overwriting output5.txt


In [36]:
%%writefile input6.txt
17
156999 342 1 432 6800 256333 300000 2345 8975 115699 78522 698563 356200 0 74444 895 9877

Overwriting input6.txt


In [37]:
698563 * 356200

248828140600

In [38]:
%%writefile output6.txt
248828140600

Overwriting output6.txt


In [39]:
%%writefile input7.txt
3
200000 8 3

Overwriting input7.txt


In [40]:
200000 * 8

1600000

In [41]:
%%writefile output7.txt
1600000

Overwriting output7.txt


In [42]:
%%writefile input8.txt
21
33562 9855 15699 41563 12566 85655 21566 10000 30000 23666 32568 14895 58966 13200 34209 10003 22605 5698 236 568 0

Overwriting input8.txt


In [43]:
85655 * 58966

5050732730

In [44]:
%%writefile output8.txt
5050732730

Overwriting output8.txt


In [25]:
%%writefile input9.txt
40
43937 41629 84392 35085 32304 78497 34755 29412 60549 37880
74134 29215 67040 12767 49848 61392 82606 17004 76380 49181
48438 55699 64113 64060 80455 86403 17854 57763 66729 25985
67521 35102 83345 21138 73794 44610 49292 56532 51801 63348

Writing input9.txt


In [26]:
86403 * 84392

7291721976

In [27]:
%%writefile output9.txt
7291721976

Writing output9.txt


In [46]:
%%writefile input10.txt
70
128071 46516 137037 69872 11039 151644 150319 102041 193016 36047
116430 134896 24591 78324 185081 81228 30564 116956 87968 175339
31607 126686 114548 79052 135702 184935 173922 35568 87968 175339
138353 92184 171429 101474 92167 83734 16674 115282 68287 132781
69614 107044 17380 176507 45097 57642 192702 153872 121837 60013
90821 39566 39140 185316 81446 140870 48057 128782 115938 115221
82174 159116 77705 139164 185212 169396 14857 187316 93554 96080

Writing input10.txt


In [47]:
193016 * 192702

37194569232

In [48]:
%%writefile output10.txt
37194569232

Writing output10.txt


In [49]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {0..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing stress_test.sh


In [ ]:
#Running tests cases. Check "/files" for "outputN.txt" where N is the Nth output for the Nth "inputN.txt" (e.g. input4.txt/output4.txt)
!chmod 744 stress_test.sh
!./stress_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [67]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [68]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [69]:
199999 * 200000

39999800000

In [70]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m55.301s
user	4m53.124s
sys	0m0.101s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m54.301s
user	4m52.421s
sys	0m0.070s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [53]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Writing max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [54]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [65]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

44
44
601



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

572


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [51]:
10 + 7 * (200000) + 6.5 * (200000)**2

260001400010.0

In [71]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177091


In [72]:
# Estimated time by instruction

295/200003177091


1.474976569326082e-09

In [73]:
# Estimated number of instructions by second

200003177091/295

677976871.4949152

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.859s
user	0m0.854s
sys	0m0.001s


## Python

In [74]:
%%writefile max_pairwise_product_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(n):
            if i != j:
                aux_product = l[i] * l[j]
                if aux_product > max_product:
                    max_product = aux_product
    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_py.py


In [75]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("11000\n")
        for i in range(1, 11001):
            f.write(f"{i} ")

Writing WriteTestPython.py


In [77]:
!python WriteTestPython.py

In [78]:
# Verify
%%time
!python max_pairwise_product_py.py < input14.txt

120989000
CPU times: user 127 ms, sys: 12.9 ms, total: 140 ms
Wall time: 19.9 s


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [79]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [80]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [81]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m15.849s
user	2m15.009s
sys	0m0.030s


In [82]:
%%writefile max_pairwise_product_n2_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(i + 1, n):
            aux_product = l[i] * l[j]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_n2_py.py


### Estimate and verify the maximun n that can be solved in 1 second



In [224]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 16400 <<"\n";  // n
 for (int i=1; i <16401; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [225]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [226]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input12.txt

^C


In [86]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("16400\n")
        for i in range(1, 16401):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [87]:
!python WriteTestPython.py

In [222]:
#Slower. Not a surprise.
%%time
!python max_pairwise_product_n2_py.py < input13.txt

268943600
CPU times: user 113 ms, sys: 16.7 ms, total: 130 ms
Wall time: 17.5 s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [89]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> v(n);                                   //n
    for(int i = 0; i < n; i++){                              //1 + 2n
        std::cin >> v[i];                                    //n
    }

    std::sort(v.begin(), v.end());                          //nlogn
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';   //2
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [90]:
%%writefile max_pairwise_product_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())                                  #1
    l = list(map(int, input().split()))               #5n
    l.sort()                                          #nlogn at least
    print(l[-1] * l[-2])                              #2

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [108]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort

!./max_pairwise_product_sort < input1.txt
!./max_pairwise_product_sort < input2.txt

90
90


In [107]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [109]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.001s


In [110]:
%%time
#Given in secs. *10^-3 ms
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 4.45 ms, sys: 688 µs, total: 5.13 ms
Wall time: 104 ms


20 stress tests

In [111]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 49 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::sort(v.begin(), v.end());
        std::cout << "n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [112]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

n: 17 Product: 850976
n: 25 Product: 857451
n: 43 Product: 980036
n: 10 Product: 692094
n: 30 Product: 839816
n: 48 Product: 956363
n: 9 Product: 475580
n: 46 Product: 954088
n: 44 Product: 902850
n: 17 Product: 981034
n: 40 Product: 928812
n: 7 Product: 267968
n: 42 Product: 991020
n: 27 Product: 908160
n: 23 Product: 990016
n: 2 Product: 141372
n: 4 Product: 418500
n: 11 Product: 557996
n: 42 Product: 963270
n: 17 Product: 690616


In [113]:
%%writefile stress_tests_sort_py.py
#Python3
import random
if __name__ == "__main__":
    for i in range(10):
        n = random.randint(4, 186)
        l = [ random.randint(27, 56734) for i in range(n) ]
        l.sort()
        print(f"n: {n} Product: {l[-1] * l[-2]}")

Writing stress_tests_sort_py.py


In [114]:
!python stress_tests_sort_py.py

n: 22 Product: 3114819890
n: 176 Product: 3208535167
n: 79 Product: 3088047120
n: 17 Product: 2568598890
n: 170 Product: 3127928206
n: 101 Product: 3018574348
n: 161 Product: 3128169177
n: 158 Product: 3180734080
n: 58 Product: 3172280073
n: 172 Product: 3195407728


Count executed instructions in code and verify with inputs 1,2 and 3


In [115]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                                                   //1
    ins++;
    std::cin >> n;                                           //1
    ins++;
    std::vector<uint64_t> v(n);                                   //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){                              //1 + 2n
        ins += 2;
        std::cin >> v[i];                                    //n
        ins++;
    }

    std::sort(v.begin(), v.end());                           //nlogn at least (nlogn time complexity)
    ins += n * log2(n);
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';    //3
    ins += 3;
    std::cout << "Executed instructions: " << ins << '\n';   // 7 + 4n + ins(nlogn sort). Where ins(nlogn sort) are, at least, nlogn.
    return 0;
}

Writing max_pairwise_product_sort_ins.cpp


In [116]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 79
90
Executed instructions: 79
40000000000
Executed instructions: 45


In [221]:
!time ./max_pairwise_product_sort_ins < input10.txt

37194569232
Executed instructions: 715

real	0m0.007s
user	0m0.001s
sys	0m0.002s


In [118]:
%%writefile max_pairwise_product_sort_ins.py
#Python3
import math

if __name__ == "__main__":
    ins = 0
    n = int(input())                       #1
    ins += 1
    l = list(map(int, input().split()))    #5n (n input read, n mappings to int, n iterable to list, n comparisons, n splits through input)
    ins += 5 * n
    l.sort()                               #nlogn at least
    ins += n * math.log2(n)
    print(l[-1] * l[-2])             #2
    ins += 2
    print(f"Executed instructions: {int(ins)}\n")

Writing max_pairwise_product_sort_ins.py


In [119]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Executed instructions: 86

90
Executed instructions: 86

40000000000
Executed instructions: 48



In [218]:
%%time
!python max_pairwise_product_sort_ins.py < input9.txt

6624518824
Executed instructions: 415

CPU times: user 5.14 ms, sys: 117 µs, total: 5.25 ms
Wall time: 105 ms


Calculate estimated time by instruction and number of instructions by second

In [220]:
estimated_time = 0.00525 / 415
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

1.2650602409638555e-05
79047.61904761904


### Estimate and verify the maximun n that can be solved in 1 second

In [122]:
# Verify input13.txt
%%writefile WriteTest.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 8000000 <<"\n";  // n
 for (int i=1; i <8000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [123]:
!g++ WriteTest.cpp -o WriteTest

In [124]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input13.txt

268943600

real	0m0.018s
user	0m0.015s
sys	0m0.000s


In [125]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input16.txt", "w") as f:
        f.write("8000000\n")
        for i in range(1, 8000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [126]:
!python WriteTestPython.py

In [237]:
%%time
!python max_pairwise_product_sort.py < input7.txt

1600000
CPU times: user 3.2 ms, sys: 1.78 ms, total: 4.98 ms
Wall time: 105 ms


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [128]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    long long p_max = 0;                     //1
    long long s_max = 0;                     //1
    for(int i = 0; i < n; ++i) {             //1 + 2n
        if (v[i] > s_max){                    //n
            s_max = v[i];                    //[0, n]
        }
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            p_max = v[i];                    //[0, n]
        }
    }
    return (long long) p_max * s_max;       //2 -> 9 + 4n + 3[0, n]           //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    std::cin >> n;                          //1
    std::vector<int> v(n);                  //n
    for(int i = 0; i < n; i++){             //1 + 2n
        std::cin >> v[i];                   //n
    }
    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n
    return 0;
}

Writing max_pairwise_product_ef.cpp


In [129]:
%%writefile max_pairwise_product_ef.py
#Python3
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product_f(l))

Writing max_pairwise_product_ef.py


Test with inputs 1 and 2



In [130]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [131]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

90
90


Test and time it with input 3

In [132]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.000s
sys	0m0.002s


In [208]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input12.txt

43145823035422802

real	1m36.653s
user	1m34.291s
sys	0m1.473s


In [134]:
%%time
!python max_pairwise_product_ef.py < input3.txt

40000000000
CPU times: user 4.93 ms, sys: 1.75 ms, total: 6.68 ms
Wall time: 109 ms


In [209]:
%%time
!python max_pairwise_product_ef.py < input12.txt

/bin/bash: line 1: 46773 Killed                  python max_pairwise_product_ef.py < input12.txt
CPU times: user 117 ms, sys: 17.9 ms, total: 135 ms
Wall time: 16.4 s


20 stress tests

In [210]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();
    ins++;
    long long p_max = 0;
    ins++;
    long long s_max = 0;
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {
        ins += 2;
        ins++;
        if (v[i] > s_max){
            s_max = v[i];
            ins++;
        }
        ins++;
        if (v[i] > p_max){
            s_max = p_max;
            ins++;
            p_max = v[i];
            ins++;
        }
    }

    return (long long) p_max * s_max;
}

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 83 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::cout << "n: " << n << " Product: " << max_pairwise_product_f(v) << '\n';
    }
    return 0;
}

Overwriting stress_tests_ef_cpp.cpp


In [211]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 27 Product: 850976
n: 59 Product: 980036
n: 80 Product: 903178
n: 65 Product: 987022
n: 45 Product: 902850
n: 45 Product: 981034
n: 47 Product: 982056
n: 3 Product: 734424
n: 76 Product: 993010
n: 63 Product: 988020
n: 25 Product: 968112
n: 51 Product: 960319
n: 55 Product: 951588
n: 9 Product: 869176
n: 6 Product: 575391
n: 33 Product: 979080
n: 2 Product: 82476
n: 40 Product: 935085
n: 4 Product: 829719
n: 83 Product: 979098


In [138]:
%%writefile stress_tests_ef_py.py
#Python3
import random
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        l = [random.randint(17, 24531) for i in range(n)]
        print(f"n: {n} Product: {max_pairwise_product_f(l)}")

Writing stress_tests_ef_py.py


In [139]:
!python stress_tests_ef_py.py

n: 60 Product: 597092790
n: 68 Product: 576872205
n: 95 Product: 565527861
n: 68 Product: 584569752
n: 20 Product: 510458586
n: 53 Product: 550897494
n: 9 Product: 441008946
n: 67 Product: 565324993
n: 91 Product: 573019202
n: 66 Product: 594052496
n: 88 Product: 590076952
n: 96 Product: 591651792
n: 85 Product: 590718625
n: 54 Product: 585369801
n: 38 Product: 566095952
n: 56 Product: 577279305
n: 25 Product: 499152968
n: 87 Product: 591892845
n: 23 Product: 577266234
n: 21 Product: 356901376


Count executed instructions in code and verify with inputs 1,2 and 3


In [140]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    ins++;
    long long p_max = 0;                     //1
    ins++;
    long long s_max = 0;                     //1
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {             //1 + 2n
        ins += 2;
        ins++;
        if (v[i] > s_max){                    //n
            s_max = v[i];                     //[0, n]
            ins++;
        }
        ins++;
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            ins++;
            p_max = v[i];                    //[0, n]
            ins++;
        }
    }
    ins += 2;
    return (long long) p_max * s_max;       //2 -> 6 + 4n + 3[0, n]          //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    ins++;
    std::cin >> n;                          //1
    ins++;
    std::vector<int> v(n);                  //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){             //1 + 2n
        ins += 2;
        std::cin >> v[i];                   //n
        ins++;
    }

    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n + 6 + 4n + 3[0, n] -> 10 +8n + 3[0, n] -> 10 + 11n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_ef_ins.cpp


In [141]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 119
90
Executed instructions: 93
40000000000
Executed instructions: 68


In [207]:
!time ./max_pairwise_product_ef_ins < input14.txt

43145823035422802
Executed instructions: 2284873006

real	1m34.668s
user	1m32.217s
sys	0m1.427s


In [143]:
%%writefile max_pairwise_product_ef_ins_py.py
#Python3
ins = 0
def max_pairwise_product_f(l):
    global ins
    p_max = 0                               #1
    ins += 1
    s_max = 0                               #1
    ins += 1
    ins += 1
    for i in l:                             #1 + 2n
        ins += 2
        if i > s_max:                       #n
            s_max = i                       #[0, n]           #[0, n] means that, in the worst case, executes n times
            ins += 1
        if i > p_max:                       #n
            s_max = p_max                   #[0, n]
            ins += 1
            p_max = i                       #[0, n]
            ins += 1

    ins += 2
    return p_max * s_max                    #2       #5 + 4n + 3[0, n] -> 5 + 7n worst

if __name__ == "__main__":
        n = int(input())                                 #1
        ins += 1
        l = list(map(int, input().split()))              #5n
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))                 #1
        print(f"Executed instructions: {ins}")           #-> 2 + 5n + 5 + 7n -> 7 + 12n

Writing max_pairwise_product_ef_ins_py.py


In [144]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


In [145]:
%%time
!python max_pairwise_product_ef_ins_py.py < input3.txt

40000000000
Executed instructions: 60
CPU times: user 5.88 ms, sys: 0 ns, total: 5.88 ms
Wall time: 105 ms


In [235]:
%%time
!python max_pairwise_product_ef_ins_py.py < input13.txt

268943600
Executed instructions: 164007
CPU times: user 4.54 ms, sys: 1.09 ms, total: 5.63 ms
Wall time: 105 ms


Calculate estimated time by instruction and number of instructions by second

In [236]:
estimated_time = 0.00563/ 164007
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

3.4327803081575786e-08
29130905.861456484


### Estimate and verify the maximun n that can be solved in 1 second

In [148]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 60000000 <<"\n";  // n
 for (int i=1; i <60000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [149]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [150]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

In [151]:
!time ./max_pairwise_product_ef_cpp < input15.txt

3599999940000000

real	0m25.626s
user	0m25.136s
sys	0m0.285s


In [154]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input15.txt", "w") as f:
        f.write("60000000\n")
        for i in range(1, 60000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [155]:
!python WriteTestPython.py

In [156]:
%%time
!python max_pairwise_product_ef.py < input15.txt

3599999940000000
CPU times: user 152 ms, sys: 20.7 ms, total: 173 ms
Wall time: 22.8 s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [157]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    int n;
    std::cin >> n;
    int p_max = INT_MIN;
    int s_max = INT_MIN;

    for (int i = 0; i < n; i++){
        int x;
        std::cin >> x;
        if(x > p_max){
            s_max = p_max;
            p_max = x;
        }
        else if(x > s_max){
            s_max = x;
        }
    }

    std::cout << (long long) p_max * s_max << "\n";
    return 0;
}

Writing max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [158]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [159]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.002s
user	0m0.001s
sys	0m0.001s


In [205]:
!time ./max_pairwise_product_gef_cpp < input13.txt

268943600

real	0m0.018s
user	0m0.010s
sys	0m0.001s


20 stress tests

In [161]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 77 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

        for (int i = 0; i < n; i++){
            int x = rand() % 1000 + 2;
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Writing stress_tests_gef_cpp.cpp


In [162]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 52 Product: 863968
n: 59 Product: 956084
n: 53 Product: 898343
n: 22 Product: 927682
n: 2 Product: 56643
n: 32 Product: 928140
n: 48 Product: 885256
n: 26 Product: 981034
n: 40 Product: 928812
n: 23 Product: 876600
n: 55 Product: 991020
n: 27 Product: 990016
n: 11 Product: 557996
n: 70 Product: 994008
n: 53 Product: 961308
n: 58 Product: 926316
n: 39 Product: 979080
n: 3 Product: 290360
n: 23 Product: 876015
n: 26 Product: 874290


Count executed instructions in code and verify with inputs 1,2 and 3

In [163]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                          //1
    ins++;
    std::cin >> n;                  //1
    ins++;
    int p_max = INT_MIN;            //1
    ins++;
    int s_max = INT_MIN;            //1
    ins++;
    ins++;
    for (int i = 0; i < n; i++){    //1 + 2n
        ins += 2;
        int x;                      //n
        ins++;
        std::cin >> x;              //n
        ins++;
        ins += 2; //both ifs
        if(x > p_max){              //[0, n]           //[0, n] means that, in the worst case, executes n times.
            s_max = p_max;          //[0, n]
            ins++;
            p_max = x;              //[0, n]
            ins++;
        }
        else if(x > s_max){         //[0, n]
            s_max = x;              //[0, n]
            ins++;
        }
    }

    ins += 3;
    std::cout << (long long) p_max * s_max << "\n"; //3 -> 8 + 4n + 5[0, n] -> 8 + 9n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_gef_ins_cpp.cpp


In [164]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!./max_pairwise_product_gef_ins_cpp < input1.txt
!./max_pairwise_product_gef_ins_cpp < input2.txt
!./max_pairwise_product_gef_ins_cpp < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52


In [233]:
!time ./max_pairwise_product_gef_ins_cpp < input14.txt

43145823035422802
Executed instructions: 1661725824

real	1m30.142s
user	1m28.541s
sys	0m0.977s


Calculate estimated time by instruction and number of instructions by second

In [234]:
estimated_time = 5408.52 / 1661725824
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

3.254760756489273e-06
307242.24445874285


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [168]:
%%writefile max_pairwise_product_gef_py.py
#Python3
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        x = int(i)
        if(x > p_max):
            s_max = p_max
            p_max = x
        elif x > s_max:
            s_max = x
    print(p_max * s_max)

Writing max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [169]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [170]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 4.6 ms, sys: 968 µs, total: 5.56 ms
Wall time: 107 ms


In [204]:
%%time
!python max_pairwise_product_gef_py.py < input10.txt

29269718304
CPU times: user 5.5 ms, sys: 0 ns, total: 5.5 ms
Wall time: 104 ms


20 stress tests

In [202]:
%%writefile stress_tests_gef_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        p_max = -1
        s_max = -1
        for x in range(n):
            i = random.randint(1, 150000)
            if(i > p_max):
                s_max = p_max
                p_max = i
            elif i > s_max:
                s_max = i
        print(f"n: {n} Product: {p_max * s_max}")

Overwriting stress_tests_gef_py.py


In [203]:
!python stress_tests_gef_py.py

n: 65 Product: 21838528976
n: 77 Product: 20308594448
n: 4 Product: 10125848310
n: 17 Product: 21320627735
n: 86 Product: 21429888636
n: 9 Product: 20429431071
n: 17 Product: 20991439054
n: 54 Product: 21892356397
n: 22 Product: 20001138000
n: 57 Product: 21408692366
n: 73 Product: 21170854188
n: 21 Product: 21974688367
n: 11 Product: 17686265500
n: 11 Product: 20173070799
n: 33 Product: 21040784122
n: 68 Product: 22202935062
n: 7 Product: 17863873552
n: 11 Product: 20582791716
n: 85 Product: 21577834191
n: 25 Product: 21216066954


Count executed instructions in code and verify with inputs 1,2 and 3

In [174]:
%%writefile max_pairwise_product_gef_ins_py.py
import random
#Python3
if __name__ == "__main__":
        ins = 0
        n = int(input())                 #1
        ins += 1
        p_max = -1                       #1
        ins += 1
        s_max = -1                       #1
        ins += 1
        ins += 1
        for i in input().split():        #1 + 3n (n reads, 2n comparisons unitl end of input)
            ins += 2
            i = int(i)                   #n
            ins += 1
            ins += 2
            if(i > p_max):               #[0, n]
                s_max = p_max            #[0, n]
                ins += 1
                p_max = i                #[0, n]
                ins += 1
            elif i > s_max:              #[0, n]
                s_max = i                #[0, n]
                ins += 1
        print(p_max * s_max)             #2
        print(f"Executed instructions: {ins}") #->6 + 4n + 5[0, n] -> 6 + 9n worst case

Writing max_pairwise_product_gef_ins_py.py


In [175]:
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 74
90
Executed instructions: 57
40000000000
Executed instructions: 42


In [231]:
%%time
!python max_pairwise_product_gef_ins_py.py < input13.txt

268943600
Executed instructions: 114804
CPU times: user 4.64 ms, sys: 0 ns, total: 4.64 ms
Wall time: 204 ms


Calculate estimated time by instruction and number of instructions by second

In [232]:
estimated_time = 0.00464 / 114804
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.041671021915613e-08
24742241.379310343


### Estimate and verify the maximun n that can be solved in 1 second

In [178]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 207715727.00296736 <<"\n";  // n
 for (int i=1; i <207715727.00296736; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [179]:
!g++ WriteTest.cpp -o WriteTest

In [180]:
!./WriteTest

In [181]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input15.txt

0

real	0m0.003s
user	0m0.001s
sys	0m0.001s


In [197]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input12.txt", "w") as f:
        f.write("207715727\n")
        for i in range(1, 207715728):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [198]:
!python WriteTestPython.py

In [199]:
%%time
!python max_pairwise_product_gef_py.py < input12.txt

/bin/bash: line 1: 43531 Killed                  python max_pairwise_product_gef_py.py < input12.txt
CPU times: user 177 ms, sys: 37.4 ms, total: 215 ms
Wall time: 18.3 s


## Perform 100 randomized automated tests in C++ and Python between the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n ln g))$.

In [195]:
%%writefile stress_tests_sort_hundred_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 100; i++){
        int n = rand() % 83 + 2;
        std::vector<int64_t> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 10000 + 2;
        }

        std::sort(v.begin(), v.end());
        std::cout << "Test: " << i + 1 << " n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }

    return 0;
}

Overwriting stress_tests_sort_hundred_cpp.cpp


In [196]:
!g++ stress_tests_sort_hundred_cpp.cpp -o hundred_cpp
!./hundred_cpp

Test: 1 n: 27 Product: 79740408
Test: 2 n: 59 Product: 97628232
Test: 3 n: 80 Product: 97780362
Test: 4 n: 65 Product: 97811259
Test: 5 n: 45 Product: 88612926
Test: 6 n: 45 Product: 94359860
Test: 7 n: 47 Product: 97207578
Test: 8 n: 3 Product: 64238964
Test: 9 n: 76 Product: 98703600
Test: 10 n: 63 Product: 87688304
Test: 11 n: 25 Product: 88777728
Test: 12 n: 51 Product: 91131372
Test: 13 n: 55 Product: 97712784
Test: 14 n: 9 Product: 96760793
Test: 15 n: 6 Product: 44497291
Test: 16 n: 33 Product: 80568215
Test: 17 n: 2 Product: 82476
Test: 18 n: 40 Product: 95000409
Test: 19 n: 4 Product: 71915366
Test: 20 n: 83 Product: 97859750
Test: 21 n: 40 Product: 88783476
Test: 22 n: 6 Product: 62983584
Test: 23 n: 17 Product: 99349934
Test: 24 n: 37 Product: 88629060
Test: 25 n: 76 Product: 92822030
Test: 26 n: 28 Product: 93790707
Test: 27 n: 59 Product: 95145141
Test: 28 n: 55 Product: 96772320
Test: 29 n: 55 Product: 95871165
Test: 30 n: 84 Product: 97494027
Test: 31 n: 73 Product: 9608

In [193]:
%%writefile stress_tests_sort_hundred_py.py
#Python3
import random
if __name__ == "__main__":
    for i in range(100):
        n = random.randint(1, 200)
        l = [ random.randint(10, 60000) for i in range(n) ]
        l.sort()
        print(f"Test: {i + 1} n: {n} Product: {l[-1] * l[-2]}")

Overwriting stress_tests_sort_hundred_py.py


In [194]:
!python stress_tests_sort_hundred_py.py

Test: 1 n: 159 Product: 3516712804
Test: 2 n: 185 Product: 3572077648
Test: 3 n: 134 Product: 3595141638
Test: 4 n: 177 Product: 3506048400
Test: 5 n: 116 Product: 3557462220
Test: 6 n: 92 Product: 3510206993
Test: 7 n: 45 Product: 3471923364
Test: 8 n: 82 Product: 3520582834
Test: 9 n: 152 Product: 3549820820
Test: 10 n: 17 Product: 3058071364
Test: 11 n: 83 Product: 3486932588
Test: 12 n: 16 Product: 2993183776
Test: 13 n: 102 Product: 3438824139
Test: 14 n: 56 Product: 3571395876
Test: 15 n: 129 Product: 3577813200
Test: 16 n: 124 Product: 3501321838
Test: 17 n: 59 Product: 3366517973
Test: 18 n: 192 Product: 3523490872
Test: 19 n: 55 Product: 3409978960
Test: 20 n: 45 Product: 3474570780
Test: 21 n: 145 Product: 3556125252
Test: 22 n: 97 Product: 3507064659
Test: 23 n: 160 Product: 3560502778
Test: 24 n: 148 Product: 3545607156
Test: 25 n: 113 Product: 3475218465
Test: 26 n: 36 Product: 3547468621
Test: 27 n: 46 Product: 3328696340
Test: 28 n: 65 Product: 3223086679
Test: 29 n: 151